packages/firebase_analytics/firebase_analytics_platform_interface/lib/src/event_parameters.dart

```diff
@@ -0,0 +1,43 @@
+const String _exceptionMessage =
+    "'string' OR 'number' must be set as the value of the parameter";
+
+class EventParameters {
```

Comment:
Isn't it, again, error-prone?

API-wise, the user is able to provide both `String` and `num`. Implementation-wise we throw an exception for a behavior already allowed by its API.

I was thinking about something like..

```dart
abstract class EventParameter {
  const factory EventParameter.bool(bool value) = EventParameterBool;
  const factory EventParameter.string(String value) = EventParameterString;
  const factory EventParameter.num(num value) = EventParameterNum;
  const factory EventParameter.map(Map<String, EventParameter> value) = EventParameterMap;

  Object get value;

  T when<T>(
    T Function(bool value) bool,
    T Function(String value) string,
    T Function(num value) num,
    T Function(Map<String, EventParameter> value) map,
  );
}

class EventParameterBool implements EventParameter {
  const EventParameterBool(this.value);

  @override
  final bool value;

  @override
  String toString() => 'EventParameterBool($value)';

  @override
  T when<T>(
    T Function(bool value) bool,
    T Function(String value) string,
    T Function(num value) num,
    T Function(Map<String, EventParameter> value) map,
  ) {
    return bool(value);
  }
}

class EventParameterString implements EventParameter {
  const EventParameterString(this.value);

  @override
  final String value;

  @override
  String toString() => 'EventParameterString($value)';

  @override
  T when<T>(
    T Function(bool value) bool,
    T Function(String value) string,
    T Function(num value) num,
    T Function(Map<String, EventParameter> value) map,
  ) {
    return string(value);
  }
}

class EventParameterNum implements EventParameter {
  const EventParameterNum(this.value);

  @override
  final num value;

  @override
  String toString() => 'EventParameterNum($value)';

  @override
  T when<T>(
    T Function(bool value) bool,
    T Function(String value) string,
    T Function(num value) num,
    T Function(Map<String, EventParameter> value) map,
  ) {
    return num(value);
  }
}

class EventParameterMap implements EventParameter {
  const EventParameterMap(this.value);

  @override
  final Map<String, EventParameter> value;

  @override
  String toString() => 'EventParameterMap($value)';

  @override
  T when<T>(
    T Function(bool value) bool,
    T Function(String value) string,
    T Function(num value) num,
    T Function(Map<String, EventParameter> value) map,
  ) {
    return map(value);
  }
}
```

Reply:
Also, the `logEvent` API can have a `Map<String, CustomEventParameter> parameters` or something like, so we would not introduce a new object which just contains a `Map` and `asMap`.

Reply:
Also, funny enough, if you think good we can provide an _iterable_ constructor, encoding to / decoding from a `String` internally. 😅

Reply:
Hey @victor-tinoco, thanks for your update. I'll be updating original implementation as mentioned [here](https://github.com/firebase/flutterfire/pull/9520#discussion_r970711696)